oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,492 @@
+OEP-65: Frontend Composability
+##############################
+
+.. list-table::
+   :widths: 25 75
+
+   * - OEP
+     - :doc:`OEP-65 <oep-0065-arch-frontend-composability>`
+   * - Title
+     - Frontend Composability
+   * - Last Modified
+     - 2024-04-12
+   * - Authors
+     -
+       * Adolfo R. Brandes <adolfo@axim.org>
+       * Pedro Martello <pedro@hammerlabs.net>
+       * David Joy <david@joy.engineering>
+   * - Arbiter
+     - Adam Stankiewicz <astankiewicz@2u.com>
+   * - Status
+     - Under Review
+   * - Type
+     - Architecture
+   * - Created
+     - 2024-04-03
+   * - Review Period
+     - April 15, 2024 - April 29, 2024
+   * - Resolution
+     - <links to any discussions where the final status was decided>
+   * - References
+     -
+       * `FC-0054 - Composable Micro-frontends Discovery <https://openedx.atlassian.net/wiki/spaces/COMM/pages/4063821827/FC-0054+-+Composable+Micro-frontends+Piral+Discovery>`_
+       * `FC-0007 - Modular MFE Domains Discovery <https://openedx.atlassian.net/wiki/spaces/COMM/pages/3614900241/CLOSED+FC-0007+-+Modular+MFE+Domains+Discovery>`_
+
+
+Abstract
+********
+
+This OEP proposes that Open edX micro-frontends (MFEs) adopt module federation as an approach to address inconsistencies, performance issues, and inflexibilities inherent in the current MFE architecture.
+
+Motivation
+**********
+
+Micro-frontends were originally designed to avoid some of the limitations of the monolithic ``edx-platform`` frontend, namely that otherwise independent teams were beholden to the build, test, and release lifecycle of the rest of the codebase. This dramatically slowed down the pace of frontend feature development, experimentation, and innovation.
+
+As a result, the Open edX "micro-frontend" architecture focused on creating an unopinionated, loosely coupled set of applications with the goal of enabling teams to iterate quickly and independently. This goal was successful.
+
+However, now roughly 6 years into this gradual re-platforming, we've discovered that the completely siloed nature of these MFEs has created its own set of problems.  Our frontends are more like independent single-page applications than micro-frontends, as we never invested in ways of composing them together. The problems inherent in today's MFE architecture are described below.
+
+Consistency
```

Comment:
Thing is, frontend composability isn't going to address places where someone has chosen to make an MFE-specific component or follow some UX pattern not used elsewhere.  That's just a choice they made which we should have caught in some review process.

It can help with consistency between dependency versions, however, which is why this section focuses on the different kinds of dependency inconsistency.  I'm trying hard to keep the Motivation section limited to stuff we're actually addressing with these changes so it doesn't become a laundry list of our MFE gripes. 😄

Reply:
> frontend composability isn't going to address places where someone has chosen to make an MFE-specific component

In the case of the header I think it does. Having the header and footer live on the shell instead of in the MFE discourages that type of change

> I'm trying hard to keep the Motivation section limited to stuff we're actually addressing with these changes so it doesn't become a laundry list of our MFE gripes. 😄

:100:

Reply:
I'm with ya.  I've reworked this section a smidge to point out that inconsistency can result from varying versions, but also because developers have more leeway to make divergent choices, and then added examples in the sub-sections.